### PR TITLE
OpenSlides history mode

### PR DESCRIPTION
**Pull request by normanjaeckel**

### CODE REVIEW

**Comment by normanjaeckel** [comment]
This is a first proof of concept. I can not test history information.

@tsiegleauq: Can you provide a set_state view please (changing state of a motion without using the full edit mode)?

@FinnStutzenstein Please have a look and tell me if this is the way we can go. The concept of KeyFrames may follow later.

**Comment by tsiegleauq** [comment]
> Can you provide a set_state view please (changing state of a motion without using the full edit mode)?

Will this be required permanently or for testing purposes only?

Would a mat-menu in the detail view be sufficient (not in edit mode)?

**Comment by normanjaeckel** [comment]
@FinnStutzenstein You can test this branch with a fresh db and try a first implementation.

Create a motion and update text and change state. Then see the new URLs for history information and history data. You can fire GET requests to get history information with timestamps and texts and also all data.

Later I will change the history information view (with timestamps and element ids and information text) back to a usual autoupdate model and split out the full data history to a separate model.

Please ask @emanuelschuetze for any questions.

**Comment by tsiegleauq** [comment]
@FinnStutzenstein I could provide a history (list)view later that day.

**Comment by FinnStutzenstein** [comment]
@tsiegleauq OK, please keep it simple first.

**Comment by FinnStutzenstein** [comment]
@normanjaeckel Django default permissions are created. Please prevent this.

**Comment by tsiegleauq** [comment]
rebased and force pushed after recommendation to do so by @FinnStutzenstein 
This is now on current master and provides a simple list to see the history:

localhost_4200_history

**Comment by emanuelschuetze** [comment]
please sort the list by date (desc)

**Comment by emanuelschuetze** [comment]
Added button to print full data output in console - for each history entry.

**Comment by normanjaeckel** [comment]
Rebase and refactored. History is now a root rest element. Please add resource core/history to the client. (I can not test at the moment).

Still open:
- Restrict access to admins only.
- Clear history view.
- Probably some bugs ...

File: openslides/core/urls.py [hunk 1]
```diff
@@ -12,11 +12,7 @@
...
 
-    url(r'^history/data/$',
-        views.HistoryDataView.as_view(),
-        name='core_history_data'),
-
-    url(r'^history/information/$',
-        views.HistoryInformationView.as_view(),
-        name='core_history_information'),
+    url(r'^history/$',
+        views.HistoryView.as_view(),
```
Comment (tsiegleauq) [inline openslides/core/urls.py]: cannot test history anymore.
On  a get requiest to `apps/core/history/`
I get: object has no attribute 'pk'

Even on a completely fresh and migrated database.

**Comment by tsiegleauq** [comment]
I integrated the root/rest history object in the client, made a history and the normal autoupdate model/viewmodel implementations.

There is a dummy-button wanting to delete everything fireing a delte request to /rest/core/history/
Feel free to make according changes

Getting a full history to a given timestamp or time string did not work for me for I cannot make proper changes.
I created the time-travel-service which is currently disabling the websocket, clearing the stores, showing a warning and preventing to send Data to the server.

It's meant to get the full history data, converting them into baseModels and to flush them into the dataStore aswell.
@normanjaeckel I suppose you can continiue from here.
Please document how to get the history to a given timestamp and ensure it's working

File: openslides/core/urls.py (hunk 1 above)
Comment (normanjaeckel) [reply openslides/core/urls.py]: Should be fixed now.

**Comment by normanjaeckel** [comment]
@tsiegleauq Please test again. Server should work now.

To trigger the clear_history please send POST to /rest/core/history/clear_history/

File: openslides/core/urls.py (hunk 1 above)
Comment (normanjaeckel) [reply openslides/core/urls.py]: Use query parameter to get filtered data:

apps/core/history/?timestamp=123456789

**Comment by tsiegleauq** [comment]
I made the time travel service functional and changed the history-page to a paginated table (so holdung 500 entries is no problem anymore)

klicking on an entry in the table will clear the datastore, disable autoupdates, cut the websocket connection and flash all data to the selected timestamp into the datastore.

Even tho i clear the storrage(service) chached data might survive. You can confirm it's working by checking the dataStore or reloading on history-the history view before selecting an entry.

**Comment by normanjaeckel** [comment]
I can not confirm the double-entries bug.

I setup a fresh db. Start the server and quit it. Then counting History objects using shell returns 105 objects.

Then I restart the server, fire a single request to /rest/core/history/clear_history/.  Then counting History objects using shell again returns 105 objects.

Of course the ids go up but not the number of objects.

**Comment by tsiegleauq** [comment]
Ok perhaps there was something wrong in my setup. I checked out the current version of HistoryModel and it's kind of working now. Still not how I expected tho.

 - `rm personal_data/var/db.sqlite3`
 - `python manage.py migrate`
 - `python manage.py runserver`
 - Σ(History) = 105 (expected 105)
 - post to `/rest/core/history/clear_history/`
 - Σ(History) = 105 (expected 0 but maybe that's just me)
 - Upload 10 Mediafiles
 - Σ(History) = 125 (expected 115)
 - post to `/rest/core/history/clear_history/`
 - Σ(History) = 115

I checked what entries that were there and which one got deleted.
For every uploaded file, a `uses/user:id` entry was created, which got deleted after hitting clear.

Is this the intended behavior?

**Comment by normanjaeckel** [comment]
> Σ(History) = 105 (expected 0 but maybe that's just me)

Here 105 is correct because the history is cleared and immediatelly rebuild with the current state of the database. The history table must not be empty during runtime because then you can not jump back to the beginning.

> Upload 10 Mediafiles
> Σ(History) = 125 (expected 115)

Maybe uploading mediafiles calls inform_changed_data twice .

> post to /rest/core/history/clear_history/
> Σ(History) = 115

Looks good.

File: client/src/app/core/services/time-travel.service.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+export class TimeTravelService {
+    /**
+     * Determine if OpenSlides is running in History mode.
+     */
+    public inHistoryMode = false;
+
+    /**
+     * Constructs the time travel service
+     *
+     * @param http "Normal" Angular httpClient is required here, since the request must not end with "/"
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/time-travel.service.ts]: Hm, we might separate the path, query and fragments as different parameters. Would be nice if you can change this in this PR, or meight be done later (then open an issue). The HttpService should build up the correct url then. (See also analogon in the WebsocketService: There is a QueryDict defined and then builded to a complete url).

File: client/src/app/core/services/time-travel.service.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * Convert historyData to BaseModels that can be send to the
+     *
+     * @param historyData the data that should shown
+     * @returns the list of BaseModels for the DataStore
+     */
+    private historyToBaseModel(historyData: HistoryData[]): BaseModel[] {
+        const oldModels: BaseModel[] = [];
+        for (const historyObject of historyData) {
+            const collectionString = historyObject.element_id.split(":")[0]
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/time-travel.service.ts]: Use `'`

File: client/src/app/core/services/time-travel.service.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        return oldModels;
+    }
+
+    /**
+     * Clears the DataStore, the indexedDB and stops the WebSocket connection
+     */
+    private async stopTime(): Promise<void> {
+        this.webSocketService.close();
+        await this.cleanDataStore();
+        await this.storageService.clear();
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/time-travel.service.ts]: Please do not clear the storage service

File: client/src/app/site/history/services/history-repository.service.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        ) {
+        super(DS, mapperService, timeTravel, History, [User]);
+    }
+
+    /**
+     * Clients usually do not need to create a history object themselves
+     * @ignore
+     */
+    public async create(): Promise<Identifiable> {
+        return null;
```
Comment (FinnStutzenstein) [inline client/src/app/site/history/services/history-repository.service.ts]: Raise an error

File: client/src/app/site/history/services/history-repository.service.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Clients usually do not need to modify existing history objects
+     * @ignore
+     */
+    public async update(): Promise<void> {
+        return null;
```
Comment (FinnStutzenstein) [inline client/src/app/site/history/services/history-repository.service.ts]: same here

File: client/src/app/site/history/services/history-repository.service.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Sends a post-request to delete history objects
+     */
+    public async delete(): Promise<void> {
+        const restPath = `rest/core/history/clear_history/`;
+        await this.httpService.post(restPath, null);
```
Comment (FinnStutzenstein) [inline client/src/app/site/history/services/history-repository.service.ts]: Maby we should make the data attribute optional..

**Comment by FinnStutzenstein** [comment]
Note: I rebased ontop of the master.

File: client/src/app/core/services/query-params.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    [key: string]: QueryParamValue;
+}
+
+/**
+ * Formats query params for the url.
+ *
+ * @param queryParams
+ * @returns the formatted query params as string
+ */
+export function formatQueryParams(queryParams: QueryParams = {}): string {
```
Comment (tsiegleauq) [inline client/src/app/core/services/query-params.ts]: please move int /core/

File: client/src/app/core/services/time-travel.service.ts [hunk 9]
```diff
@@ -50,10 +46,11 @@ export class TimeTravelService {
      */
     public constructor(
         private http: HttpClient,
-        private storageService: StorageService,
         private webSocketService: WebsocketService,
         private modelMapperService: CollectionStringModelMapperService,
-        private DS: DataStoreService
+        private DS: DataStoreService,
```
Comment (tsiegleauq) [inline client/src/app/core/services/time-travel.service.ts]: ensure comments (dont see em now)

**Comment by FinnStutzenstein** [comment]
@normanjaeckel Python tests are failing

**Comment by tsiegleauq** [comment]
Rebased to current master, and meld the last 3 commits together (to save steps during potential future rebases).
@FinnStutzenstein masde some adjustments to time traveling to syncronize it better with the autoUpdate/datastore routine.

The client side works well.

**Comment by normanjaeckel** [comment]
Server tests are ok now.

**Comment by FinnStutzenstein** [comment]
@tsiegleauq Can you lock down typescript?

**Comment by tsiegleauq** [comment]
rebasing to current master now after #4044

**Comment by tsiegleauq** [comment]
one open todo to display the table a little nicer. I am on it.

**Comment by FinnStutzenstein** [comment]
@tsiegleauq You may accumulate by timestamp. So just giving each timestamp once and then (maybe as "quickedit" tables) list all changed items. So having a timestamp x with (motion 3, user 4, modelx, id y).

**Comment by tsiegleauq** [comment]
@FinnStutzenstein in another pull request perhaps. Ideally the server would do that for us.

**Pull request merged**